### PR TITLE
Rework command interpreters

### PR DESCRIPTION
**Pull request by rliebz**
Rework command interpreters

Using the `SHELL` environment variable to determine the command
interpreter was problematic because the variable signals the shell that
is being used, not the shell that is intended to run scripts in a
project. That said, for systems that do not have have an executable on
their path named `sh`, or if `sh` is less desirable, there needs to be a
way to specify what should be used instead.

The solution here is to allow users to specify their own interpreter and
the flags needed to invoke it. This effectively removes the requirement
for an interpreter to match the following pattern:

    ${SHELL} -c "..."

The range of supported interpreters is therefore expanded to include
anything that can run an arbitrary script passed as a string, such as
node or ruby, whether or not they happen to have a `-c` flag.

This setting may only be specified at the file level rather than the
task level. While there may be semantics that make sense to allow for
more granular specification of the interpreter, those are not currently
clear to me, and I don't necessarily know that they are worth
supporting.

This change also unifies commands run during `when` or `option` clauses
so that they respect the interpreter setting the same way that a `command`
does.

Resolves: #72

### CODE REVIEW

File: runner/metadata_test.go [hunk 1]
```diff
@@ -252,6 +252,91 @@ func TestMetadata_Set(t *testing.T) {
...
+		{
+			name: "env",
+			env:  `/usr/bin/env zsh`,
+			want: []string{"/usr/bin/env", "zsh"},
+		},
+		{
+			name:   "config file and env",
+			config: `bash`,
+			env:    `/usr/bin/env zsh`,
+			want:   []string{"/usr/bin/env", "zsh"},
```
Comment (smyrman) [inline runner/metadata_test.go:280]: Is this a typo? (`config: bash`), want `zsh`...

File: docs/spec.md [hunk 2]
```diff
@@ -618,6 +620,32 @@ It is invalid to split the configuration; if the `include` clause is used, no
...
+
+```text
+<interpreter> [optional-arg]
+```
+
+Unlike a shebang, however, it is not a requirement that the interpreter is the
+absolute path to the executable, so long as it is available on the `PATH`.
+
+This value can also be using the `TUSK_INTERPRETER` environment variable. The
+environment variable has higher precidence than the configuration value.
```
Comment (smyrman) [inline docs/spec.md:647]: I think it should be opposite; a configuration inside the file _should_ override the environment. Thus if setting `TUSK_INTERPETER` to `zsh`, we won't try to use `zsh` to run a tusk file with `interpeter` set to `python`. 

Alternatively, perhaps there is no strong need for an environment variable at all. While it works OK now, it may make it even harder to introduce per-task interpreter settings later?

Doesn't matter too much to me though, if you think it's important to be able to override, thats _fine_. I would rather just edit the file, keeping it declarative.

File: runner/command.go [hunk 3]
```diff
@@ -51,10 +57,46 @@ func (c *Command) UnmarshalYAML(unmarshal func(interface{}) error) error {
...
+//
+// A cleanup function is returned to allow removal of the temporary file.
+func newCmd(ctx Context, script string) (*exec.Cmd, func(), error) {
+	file, err := ioutil.TempFile("", "tusk_exec")
+	if err != nil {
+		return nil, nil, err
+	}
+	cleanup := func() {
+		os.Remove(file.Name()) // nolint: errcheck
+	}
```
Comment (smyrman) [inline runner/command.go:71]: While creating a temp file is a nice _fallback_, I am not sure it's the right default. E.g. there could be security reasons for not wanting to create a temp file, such as when having a plain-text password (extracted from some vault) present on the task command-line from some variable.

It might be good (in my opinion), to:
1. If the interpreter argument for running without a temp file is _known_, use that.
2. If not, fallback to creating a temporary file.
3. State it in the documentation which interpreters are that don't generate temporary files.

This way anyone that has a security concern in relation to creating temporary files can take the right precaution.

I am far from an security expert, so I do not know if this makes any sense... Just a thought.

File: docs/spec.md (hunk 2 above)
Comment (rliebz) [reply docs/spec.md:647]: > Alternatively, perhaps there is no strong need for an environment variable at all. 

I think being conservative here is the right call. I'll remove environment variable configuration for now, and reconsider if there's a strong case for it in the future.

File: runner/command.go (hunk 3 above)
Comment (rliebz) [reply runner/command.go:71]: It's a good thought, I'm going to have to think on this for a bit.

I lean toward being consistent over being comprehensive, so if temporary files are an issue, it probably means either supporting a whitelist of interpreters or only supporting the `${SHELL} -c "..."` pattern.

File: runner/metadata_test.go (hunk 1 above)
Comment (rliebz) [reply runner/metadata_test.go]: It was meant to test environment variable precedence, but this test case is now removed.

File: runner/command.go (hunk 3 above)
Comment (smyrman) [reply runner/command.go:71]: PS! I don't _know_ if temporary files is an issue, or _when_ they are an issue.

In the example, one could imagine shells explicitly extracting "secrets" from the environment to prevent them being included in temp files. But that requires the _knowledge_ that tusk creates temporary files, which means the risk of mistakes may be higher.

From a better safe then sorry approach, then a white-list sounds like a nice place to start. If wrting to temporary files proves not to be a real problem, going from an explicit whitelist like a write-to-file general solution would be a backwards compatible change. Going the other way would not be.

Comment (smyrman) [reply runner/command.go:71]: However, keep in mind there is a _a lot of shells_ out there... `bash`, `fish`, `dash`++

Not saying it's important to support them all.

One idea would be to allow two configuration formats...

A "quick" format (either allowing interpreters white-listed by tusk (map lookup), or simply assuming `args: ["-c"]` and `use_temp_file: false`):

```yaml
interpreter: python
```
A detailed format where you can set any interpreter you like:

```yaml
interpreter:
   command: "ruby"
   args: ["-e"]
```
```yaml
interpreter:
   command: "gosh"
   use_temp_file: true
```

Then it becomes very obvious when you are doing what.

Comment (smyrman) [reply runner/command.go:71]: Again, this is something that can be added _later_ of starting out with a white-list.

Comment (rliebz) [reply runner/command.go:71]: After thinking and talking it over with a couple folks, I don't think using a temporary file is worth the trade-offs here. The difference is only that we can't look like shebangs, which is superficial, and we can't support interpreters that can't take a literal string as a flag but can take a file, which doesn't describe any scripting language I can think of.

What's making sense to me is a configurable option that defaults to `sh -c`, but makes no assumptions about `-c` or other flags. Specified with one of these syntaxes, most likely:

1. ```yaml
    interpreter: python -c
    ```

1. ```yaml
    interpreter: [python, -c]
    ```
1. ```yaml
    interpreter:
      executable: python # maybe `path:`, maybe `command:`
      args: [-c]
    ```

We can always do something like start with option 1 or 2, then make it the short-hand version of a longer configuration like 3 down the line. It's certainly not without precedent for tusk.

A white-list that makes any assumptions (e.g., `ruby` means `ruby -e`) is something I want to stay away from. Once tusk supports it, we're stuck with a list of special interpreters until we make a breaking change.

Comment (smyrman) [reply runner/command.go:71]: Or maybe the string can be passed to STDIN? Seam to work for a number of shells without having to specify any `-c`, `-e` etc.

```sh
$ echo 'print("Hello world")' | python3                                              :(
Hello world
```
```sh
$ echo 'echo "Hello world"' | sh
Hello world
```
```sh
echo 'print("Hello world\n")' | ruby
Hello world
```

Comment (smyrman) [reply runner/command.go:71]: I don't have a set-up for it, but would be nice to find out if this can also work for `cmd.exe` (Batch) in Windows?

Comment (rliebz) [reply runner/command.go:71]: Using stdin is something I played around with briefly, but unfortunately it means you can't read the end-user's stdin. So you wouldn't be able to do things like prompt for input or open a REPL.

Comment (rliebz) [reply runner/command.go:71]: I do zero windows development, so windows support has mostly been guess work and making sure Appveyor continues to pass. I can't speak to `cmd.exe` in particular.

Comment (smyrman) [reply runner/command.go]: Good point about STDIN. From some research, `cmd.exe` is the terminal application, so never mind that! The executable for PowerShell is `PowerShell` or `PowerShell.exe`, and it seams to allow a `-Command` option.

This means your suggested solution should work, also for that shell 👍

Docs:
- https://docs.microsoft.com/en-us/powershell/scripting/components/console/powershell.exe-command-line-help?view=powershell-7
- https://docs.microsoft.com/en-us/powershell/module/Microsoft.PowerShell.Core/About/about_PowerShell_exe?view=powershell-5.1

Besides `bash` is available on new Windows environments these days.

File: runner/command_test.go [hunk 4]
```diff
@@ -69,22 +69,80 @@ func TestCommand_UnmarshalYAML(t *testing.T) {
...
 func TestCommand_exec_helper(*testing.T) {
-	if os.Getenv("TUSK_WANT_TEST_COMMAND") != "1" {
+	if os.Getenv("TUSK_TEST_EXEC_COMMAND") != "1" {
 		return
 	}
 	defer os.Exit(0)
 
 	fail := func(msg interface{}) {
-		fmt.Fprintln(os.Stderr, msg)
+		fmt.Fprintln(os.Stdout, msg)
```
Comment (smyrman) [inline runner/command_test.go:145]: Why is this changed to `Stdout`? It's good te be able to separate `Stdout` from `Stderr` so that ouput can be piped without handling log output. E.g.
```sh
tusk kubctl-secrets | kubectl apply -f -
```

**Comment by rliebz** [comment]
Thanks for the help @smyrman!
